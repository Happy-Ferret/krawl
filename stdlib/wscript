def build(bld):
	libs = 'helloworld string'.split()
	for l in libs:
		src = bld.path.ant_glob('std/%s/*.krl' % l)
		bld(
			features     = 'krawl',
			hash_uid     = "krawl standard library: %s" % l,
			source       = src,
			target       = 'std/%s' % l,
			target_brl   = 'std/%s.brl' % l,
			includes     = '.',
			package      = l,
			install_path = '${PREFIX}/include/krawl/std'
		)

	lib = ['std/%s' % l for l in libs]
	bld(
		features     = 'c cstlib',
		target       = 'stdkrawl',
		use          = lib,
		install_path = '${PREFIX}/lib/krawl'
	)
