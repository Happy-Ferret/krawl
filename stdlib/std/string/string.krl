import "stdlib.h"
import "string.h"
import "stdio.h"

type String struct {
	data *byte
	len, cap uint
}

func New(cap uint) (s *String) {
	s = stdlib.malloc(sizeof(type String))
	s.data = stdlib.malloc(cap + 1)
	s.cap = cap
	s.len = 0
}

func Free(s *String) {
	stdlib.free(s)
	stdlib.free(s.data)
}

func FromCStr(cstr *byte) (s *String) {
	s = stdlib.malloc(sizeof(type String))
	s.data = string.strdup(cstr)
	s.len = string.strlen(cstr)
	s.cap = s.len
}

func Print(s *String) {
	stdio.printf("%.*s\n", s.len, s.data);
}
