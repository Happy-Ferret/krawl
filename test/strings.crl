import "stdio.h"
import "stdlib.h"
import "string.h"

type String struct {
	data *byte
	len, cap uint
}

func StringNew(cap uint) (s *String) {
	s = stdlib.malloc(sizeof(type String))
	s.len = 0
	s.cap = cap
	s.data = stdlib.malloc(cap + 1)
	s.data[0] = 0
}

func StringFree(s *String) {
	stdlib.free(s.data)
	stdlib.free(s)
}

func StringFromCStr(cstr *byte) (s *String) {
	s = StringFromCStrLen(cstr, string.strlen(cstr).(uint))
}

func StringFromCStrLen(cstr *byte, len uint) (s *String) {
	s = stdlib.malloc(sizeof(type String))
	s.len = len
	s.cap = len
	s.data = stdlib.malloc(s.cap + 1)
	if len != 0 {
		string.memcpy(s.data, cstr, len)
	}
	s.data[len] = 0
}

func StringPrintf(fmt *byte, ...) (s *String) {
	var va va_list
	va_start(&va)
	len := stdio.vsnprintf(nil, 0, fmt, va).(uint)
	va_end(&va)

	s = stdlib.malloc(sizeof(type String))
	s.len = len
	s.cap = len
	s.data = stdlib.malloc(s.cap + 1)
	va_start(&va)
	stdio.vsnprintf(s.data, len+1, fmt, va)
	va_end(&va)
}

func main(argc int, argv **byte) int {
	s := StringNew(0)
	stdio.printf("len: %d, cap: %d, data: '%s'\n",
		     s.len, s.cap, s.data)
	StringFree(s)

	s = StringFromCStr("hello, nsf")
	stdio.printf("len: %d, cap: %d, data: '%s'\n",
		     s.len, s.cap, s.data)
	StringFree(s)

	s = StringPrintf("Hello: %d, %f", 333, 3.1415.(double))
	stdio.printf("len: %d, cap: %d, data: '%s'\n",
		     s.len, s.cap, s.data)
	StringFree(s)
	return 0
}
