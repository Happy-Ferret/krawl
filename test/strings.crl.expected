len: 0, cap: 0, data: ''
len: 10, cap: 10, data: 'hello, nsf'
len: 20, cap: 20, data: 'Hello: 333, 3.141500'
